File: swordOffer/reversePrint/reversePrint.cc
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(nullptr) {}
};

class Solution {
 public:
  vector<int> reversePrint(ListNode* head) {

    if (head == nullptr) {
      return vector<int>();
    }

    stack<ListNode*> stk;
    vector<int> res;
    for (ListNode *cur = head; cur != nullptr; cur = cur->next) {
      stk.push(cur);
    }

    while (!stk.empty()) {
      res.push_back(stk.top()->val);
      stk.pop();
    }

    return res;
  }
};

File: swordOffer/go/stackQueue/reversePrint.go
package main

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func reversePrint(head *ListNode) []int {
	if head == nil {
		return nil
	}

	var prev *ListNode
	cur := head
	for cur != nil {
		next := cur.Next
		cur.Next = prev
		prev = cur
		cur = next
	}

	var res []int
	for cur := prev; cur != nil; cur = cur.Next {
		res = append(res, cur.Val)
	}

	return res
}
